### PR TITLE
Set a descriptive name for discovered participants if none is given by the user [#11907]

### PR DESCRIPTION
**Pull request by SerranoePros**
For the review: Is better to go commit by commit.

### CODE REVIEW

File: src/cpp/subscriber/StatisticsParticipantListener.cpp [hunk 1]
```diff
@@ -238,6 +238,91 @@ StatisticsParticipantListener::StatisticsParticipantListener(
...
+{
+    // The participant_id can be obtained from the last 4 octets in the GUID prefix
+    std::stringstream buffer;
+    for (int i = 0; i < 3; i++)
+    {
+        buffer << std::hex << std::setfill('0');
+        buffer << std::setw(2) << static_cast<unsigned>(guid.guidPrefix.value[i + 8]);
+        buffer << ".";
+    }
+    buffer << std::hex << std::setfill('0');
```
Comment (JLBuenoLopez-eProsima) [inline src/cpp/subscriber/StatisticsParticipantListener.cpp:320]:
```suggestion
    buffer << std::hex << std::setfill('0');
    for (int i = 0; i < 3; i++)
    {
        buffer << std::setw(2) << static_cast<unsigned>(guid.guidPrefix.value[i + 8]);
        buffer << ".";
    }
```

It is redundant to set again the format of the buffer. Unless you set the flags with other values, the settings persist.

**Review by JLBuenoLopez-eProsima — CHANGES_REQUESTED** [review]

**Comment by JLBuenoLopez-eProsima** [comment]
The tests added in this PR are issuing `GMOCK WARNING`s. Please, try to get rid of these warnings.

File: src/cpp/subscriber/StatisticsParticipantListener.cpp [hunk 2]
```diff
@@ -238,6 +238,91 @@ StatisticsParticipantListener::StatisticsParticipantListener(
...
+    }
+
+    // 4. metatraffic_locators.multicast
+    if (search_address_in_locators(info.metatraffic_locators.multicast, address))
+    {
+        return address;
+    }
+
+    // The only option is for localhost to be the only valid IP
+    return "127.0.0.1";
```
[the remark this answers is not in the document]
Comment (SerranoePros) [reply src/cpp/subscriber/StatisticsParticipantListener.cpp:305]: I changed the return to "localhost"

File: test/unittest/StatisticsParticipantListener/StatisticsParticipantListenerTests.cpp [hunk 3]
```diff
@@ -256,6 +256,491 @@ TEST_F(statistics_participant_listener_tests, new_participant_undiscovered)
                 info)), eprosima::statistics_backend::BadParameter);
 }
 
+TEST_F(statistics_participant_listener_tests, new_participant_discovered_empty_name_no_locators)
```
[the remark this answers is not in the document]
Comment (SerranoePros) [reply test/unittest/StatisticsParticipantListener/StatisticsParticipantListenerTests.cpp:259]: I'm agree, but is not necessary a new test, I added the locators (with a valid IP) with less priority in the `new_participant_discovered_empty_name_"default/metatraffic"_"unicast/multicast"_locator` tests and check that the participant have the name of the most priority locator.

**Review by JLBuenoLopez-eProsima — COMMENTED** [review]
Some small nitpicks to clarify the comments

**Review by JLBuenoLopez-eProsima — APPROVED** [review]
:+1:

**Pull request merged**